src/issues/shareProviders.ts

```diff
@@ -0,0 +1,242 @@
+	) {
+		this.initialize();
+	}
+
+	public dispose() {
+		this.disposables.forEach((d) => d.dispose());
+		this.shareProviderRegistrations?.map((d) => d.dispose());
+	}
+
+	private async initialize() {
+		if ((await this.hasGitHubRepositories()) && this.shouldRegister()) {
+			this.register();
+		} else {
+			this.disposables.push(this.gitAPI.onDidOpenRepository(async e => {
+				await e.status();
+				if ((await this.hasGitHubRepositories()) && this.shouldRegister()) {
+					this.register();
+				}
+			}));
+		}
+		this.disposables.push(this.gitAPI.onDidCloseRepository(() => {
+			if (!this.hasGitHubRepositories()) {
+				this.unregister();
+			}
+		}));
+	}
+
+	private async hasGitHubRepositories() {
+		for (const folderManager of this.repositoryManager.folderManagers) {
+			await folderManager.repository.status();
+			if ((await folderManager.computeAllGitHubRemotes()).length) {
+				return true;
+			}
+			return false;
+		}
+	}
+
+	private register() {
+		if (this.shareProviderRegistrations) {
+			return;
+		}
+
+		this.shareProviderRegistrations = supportedSchemes.map((scheme) => vscode.window.registerShareProvider({ scheme }, this));
+	}
+
+	private unregister() {
+		this.shareProviderRegistrations?.map((d) => d.dispose());
+		this.shareProviderRegistrations = undefined;
+	}
+
+	protected abstract shouldRegister(): boolean;
+	protected abstract getBlob(folderManager: FolderRepositoryManager, uri: vscode.Uri): Promise<string>;
+	protected abstract getUpstream(repository: Repository, commit: string): Promise<Remote | undefined>;
+
+	public async provideShare(item: vscode.ShareableItem): Promise<vscode.Uri | undefined> {
+		// Get the repository that this URI is associated with
+		const repository = getRepositoryForFile(this.gitAPI, item.resourceUri);
+		if (!repository) {
+			throw new Error(vscode.l10n.t('Current file does not belong to an open repository.'));
+		}
+
+		// Get the blob
+		const folderManager = this.repositoryManager.getManagerForFile(item.resourceUri);
+		if (!folderManager) {
+			throw new Error(vscode.l10n.t('Current file does not belong to an open repository.'));
+		}
+		const blob = await this.getBlob(folderManager, item.resourceUri);
+
+		// Get the upstream
+		const remote = await this.getUpstream(repository, blob);
+		if (!remote || !remote.fetchUrl) {
+			throw new Error(vscode.l10n.t('The selection may not exist on any remote.'));
+		}
+
+		const origin = getUpstreamOrigin(remote, this.origin).replace(/\/$/, '');
+		const path = encodeURIComponentExceptSlashes(item.resourceUri.path.substring(repository.rootUri.path.length));
+		const range = getRangeSegment(item);
+
+		return vscode.Uri.parse([
+			origin,
+			'/',
+			getOwnerAndRepo(this.repositoryManager, repository, { ...remote, fetchUrl: remote.fetchUrl }),
+			'/blob/',
+			blob,
+			path,
+			range
+		].join(''));
+	}
+}
+
+export class GitHubDevShareProvider extends AbstractShareProvider implements vscode.ShareProvider {
```

Comment:
Should this handle both github.dev and vscode.dev?

I think we might need a concept of "entry point" for this API. We have a "Copy vscode.dev Link" command that exists in several places. In some of those places (editor context) it includes a path and range. In other places (file share) it does not include the path and range.

Reply:
I was hoping to abstract that away from the providers by having what each provider returns be solely dependent on the `ShareableItem` context that is passed without also needing to account for where actions are being positioned in the UI. For example, in the file/share menu we would pass a workspace folder root URI and in the editor context menu we would pass the document's full URI. It's possible that we might still need to surface a way to exclude providers from certain contributions, but I'd like to avoid that until we have a really compelling reason to add it.

Reply:
> Should this handle both github.dev and vscode.dev?

Right now RemoteHub contributes the vscode.dev links in web and the builtin GitHub extension contributes the vscode.dev links in desktop, which is a bit of a pain from an implementation standpoint to maintain different providers.  If we were to move vscode.dev links into GHPRI then users would only get those links after having both RemoteHub and GHPRI installed on VS Code desktop, so we should be careful about that if we do consolidate.